pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -576,11 +576,62 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
     {
         StablePhantomPoolUserData.JoinKindPhantom kind = userData.joinKind();
 
-        if (kind == StablePhantomPoolUserData.JoinKindPhantom.COLLECT_PROTOCOL_FEES) {
+        if (kind == StablePhantomPoolUserData.JoinKindPhantom.EXACT_TOKENS_IN_FOR_BPT_OUT) {
+            return _joinExactTokensInForBPTOut(balances, scalingFactors, protocolSwapFeePercentage, userData);
+        } else if (kind == StablePhantomPoolUserData.JoinKindPhantom.COLLECT_PROTOCOL_FEES) {
             return _collectProtocolFees();
+        } else {
+            _revert(Errors.UNHANDLED_JOIN_KIND);
         }
+    }
 
-        _revert(Errors.UNHANDLED_BY_PHANTOM_POOL);
+    function _joinExactTokensInForBPTOut(
+        uint256[] memory balances,
+        uint256[] memory scalingFactors,
+        uint256 protocolSwapFeePercentage,
+        bytes memory userData
+    )
+        private
+        returns (
+            uint256,
+            uint256[] memory,
+            uint256[] memory
+        )
+    {
+        (uint256[] memory amountsIn, uint256 minBPTAmountOut) = userData.exactTokensInForBptOut();
+        InputHelpers.ensureInputLengthMatch(_getTotalTokens() - 1, amountsIn.length);
+
+        uint256[] memory scaledAmountsInWithBpt = _addBptItem(amountsIn, 0);
+        _upscaleArray(scaledAmountsInWithBpt, scalingFactors);
+
+        // Balances come in from onJoinPool already upscaled
+        uint256 bptAmountOut = _computeBptAmountOut(balances, scaledAmountsInWithBpt);
```

Comment:
I'm not a fan of this 'compute' function - it's an odd abstranction, and doesn't do much other than hide details.

I propose an alternative refactor: we inline the 'compute' function, but create a function `_upscaleAmountsWithoutBPT` that does the whole insert-upscale-drop dance, explaining why that's required.

Reply:
It's only there because I ran into stack-too-deep. I can try it another way

Reply:
There seem to be just too many local variables need to inline it. This might be as clean as it gets; I added a lot of documentation.